### ISSUE TITLE
Allow function developers to access Kubernetes Secrets

### ISSUE THREAD

**Issue by zroubalik** [knative-sandbox/kn-plugin-func#362]
Function developers should be able to specify secrets that needs to be mounted or added as ENV variable to the deployed function.
They can do that via editing `func.yaml` file or via `func secret` command, which should interactively suggest secrets available in the namespace on the cluster and then allow users to attach them to the function.

**func.yaml example:**
```yaml

name: my-func
namespace: "test"
runtime: node
...
volumes: 
   - secret: example-secret                              # mount secret as a Volume
     path: /etc/secret-volume
   - secret: example-secret2
     path: /etc/secret-volume2
   - configMap: example-cm                               # possibility to mount a ConfigMap
     path: /etc/cm-volume
envs:
   - name: EXAMPLE1                                      # ENV directly from a value 
     value: value1
   - name: EXAMPLE2                                      # ENV from the local ENV var 
     value: {{ env.MY_ENV }}
   - name: EXAMPLE3                                      # ENV from a key in secret
     value: {{ secret.example-secret.key1 }}      
   - value: {{ secret.example-secret1 }}                 # all ENVs from secret
```

**Comment by embano1** [comment]
Thank you for the details @zroubalik. Two questions:

1) Can you provide an example with namespaces?
1) It looks like Example 3 above is incorrect (multiple value fields, incl. inconsistent use of array/non-array semantics)?

**Comment by zroubalik** [comment]
> Thank you for the details @zroubalik. Two questions:
> 
>     1. Can you provide an example with namespaces?

Do you mean referencing a Secret from a different namespace? That's not possible currently and honestly I don't think this is a good practice wrt security. 

>     2. It looks like Example 3 above is incorrect (multiple value fields, incl. inconsistent use of array/non-array semantics)?
No it is correct, but those are 2 separate examples:
This sets `key1` from the secret as a env value `EXAMPLE3`:
```yaml
envs:
   - name: EXAMPLE3
     value: {{ secret.example-secret.key1 }}        
```

This sets all key=value pairs from the secret as env variables:
```yaml
envs:
   - value: {{ secret.example-secret1 }}         
```

**Comment by embano1** [comment]
> Do you mean referencing a Secret from a different namespace? That's not possible currently and honestly I don't think this is a good practice wrt security. 

Ah sorry for the confusion here as I read the initial comment that this should be possible. I just reread and noticed my wrong assumption. I also agree that secrets should be assumed to live in the same namespace. All good then.

> No it is correct, but those are 2 separate examples:

Gotcha! Got tripped off by the missing `name` property, which of course is not missing in the examples. 👍

**Comment by lance** [comment]
@zroubalik this can be closed now that #396 has landed, correct?
